File: 101-200/q151_reverse_words.c
// =========== 2 solutions =============

//solve with spaces in start and end restriction
char * reverseWords(char * s){

    int len = strlen(s);
    char *ret = calloc(len + 1,  sizeof(char)); //ret string
    
    int ret_i = 0;
    int i = len - 1;
    
    //move backward in string stop when discovering space (spaces in the end just skipped)
    //or discovering space in start
    //move on original string backword and copy it to ret string normally
    for(; i>=0; --i)
    {
        if((i > 0 && ' ' != s[i] && s[i-1] == ' ') || 
          (0 == i && ' ' != s[i]))
        {
            //if the ret index is not 0, put a space as space between words
            if(0 != ret_i)
            {
                ret[ret_i++] = ' ';
            }

            int j = i;//where to start to copy, copy from j

            for(; ' ' != s[j] && j<len; ++j, ++ret_i)
            {
                ret[ret_i] = s[j];
            }
        }
    }
    
    return ret;
}

// ===================================
//solution without spaces in start and end restriction. - from quiz

void reverse(char *begin, char *end)
{
    while(begin < end)
    {
        char temp = *begin;
        *begin = *end;
        *end = temp;
        ++begin;
        --end;
    }
}

char *reverseWords(char *s){

    char *begin = s;
    char *runner = s;
    
    while('\0' != *runner)
    {
        if(' ' == *runner)
        {
            reverse(begin, runner - 1);
            begin = runner + 1;
        }
        ++runner;
    }
    
    reverse(begin, runner - 1);
    reverse(s, runner -1);
    
    return s;
}

File: 101-200/q151_reverse_words.java
// =========== 2 solutions =============

//no helpers at all
class Solution {
    public String reverseWords(String s) {
        
        if(null == s)
        {
            return null;
        }
        
        char []temp = s.toCharArray(); //string is immutable in java, need an extra space
        
        int n = s.length();
        
        //reverse whole string
        reverse(temp, 0, n-1);
        
        //reverse each word
        reverseWords(temp, n);
        
        //clean spaces
        return cleanSpaces(temp, n);
    }
    
    private void reverseWords(char []temp, int n)
    {
        int space_index = 0,  non_space_index = 0;
        
        while(space_index<n)
        {
            while(space_index<non_space_index || space_index < n && ' ' == temp[space_index] )
            {
                ++space_index; //skip spaces
            }
            
            while(non_space_index<space_index || non_space_index < n && ' ' != temp[non_space_index] )
            {
                ++non_space_index; //skip non spaces
            }
            
            reverse(temp, space_index, non_space_index - 1);
        }
    }
    
    private void reverse(char []temp, int start, int end)
    {
        while(start < end)
        {
            char temp_c = temp[start];
            temp[start] = temp[end];
            temp[end] = temp_c;
            
            ++start;
            --end;
        }
    }
    
    private String cleanSpaces(char []temp, int n)
    {
        int space_index = 0,  non_space_index = 0;
        
        while(space_index<n)
        {
            while(space_index < n && ' ' == temp[space_index] )
            {
                ++space_index; //skip spaces
            }
            
            while(space_index < n && ' ' != temp[space_index] )
            {
                temp[non_space_index] = temp[space_index]; //keep non spaces
                ++non_space_index;
                ++space_index;
            }
            
            while(space_index < n && ' ' == temp[space_index] )
            {
                ++space_index; //skip spaces (after word)
            }

            // keep only one space (if we have more than one space between words)
            if(space_index < n)
            {
                temp[non_space_index] = ' ';
                ++non_space_index;
            }
        }
        
        return new String(temp).substring(0, non_space_index);
    }
}

//with hepers - StringBuilder (needed) and split methods (for spaces)
public class Solution {
    public String reverseWords(String s) {
        
        String[] words = s.split(" "); //split words in spaces
        StringBuilder sb = new StringBuilder(); //string is immutable in java
        
        for(int i = 0; i<=words.length-1; ++i)
        {
            if(!words[i].isEmpty())
            {
                sb.insert(0, words[i]);
                if(i < words.length - 1)
                {
                    sb.insert(0, " ");//space after word
                }
            }
        }
        
        return sb.toString();
    }
}